Upgrade cardano-base dependency

tdammers opened the pull request:
# Description

Upgrade dependency on `cardano-base` in order to be compatible with upcoming versions of `ouroboros-consensus`.
In current versions of `cardano-base`, KES sign keys are stored in mlocked memory, and all operations involving them must now happen in `IO` or `ST`. In order to facilitate existing pure code that isn't security-critical (e.g., generating mock data for testing purposes), the previous pure KES API is now provided as `UnsoundPureKES`, which we use here.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

tdammers committed:
Upgrade cardano-base dependency

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Types.hs:
```diff
@@ -50,6 +50,7 @@ import Cardano.Ledger.Keys as X (
   KeyRole (..),
   SignKeyDSIGN,
   SignKeyKES,
+  UnsoundPureSignKeyKES,
```
Unsound stuff should not be exported from the library, because it is only suitable for testing
```suggestion
```

tdammers replied:
I'll remove it, because it doesn't really need to be exported here, but unfortunately it is not just used for testing.
The UnsoundPure API is basically the old, non-mlocked KES API; we need to keep it around for consensus to load KES sign keys from disk. This is inherently unsound, and we want to phase it out, but until we do, the unsound API is going to be used in production, just like it has been (under a less abrasive name) since the introduction of KES.

lehins replied:
> The UnsoundPure API is basically the old, non-mlocked KES API; we need to keep it around for consensus to load KES sign keys from disk.

I understand this. But there is no need to re-export this unsound serialization from other APIs if it will be removed in the near future.

tdammers replied:
Agree. Which is why I've removed it.

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Crypto.hs:
```diff
@@ -85,14 +90,25 @@ decodeVerKeyKES :: C.KESAlgorithm v => Decoder s (C.VerKeyKES v)
+encodeSignKeyKES :: (MonadST m, MonadThrow m, C.UnsoundKESAlgorithm v)
+                 => C.SignKeyKES v
+                 -> m Encoding
+encodeSignKeyKES = fmap fromPlainEncoding . C.encodeSignKeyKES
```
This as well as instances for `Enc/DecCBOR` SignKeyKES can be removed from `cardano-ledger-binary`. Make sure to bump up the major version in the package and update CHANGELOG

We do not use KES in ledger, so there is no point for us providing them. It should be up to actual users of KES (i.e. consensus) to decide which methods o serialization they want to use

lehins commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
```diff
@@ -42,7 +42,8 @@ library
         -Wall -Wcompat -Wincomplete-record-updates
-        -Wincomplete-uni-patterns -Wredundant-constraints -Wunused-packages
+        -Wincomplete-uni-patterns -Wredundant-constraints
+    -- -Wunused-packages
```
Please, don't disable any warnings. They are there for a reason
```suggestion
        -Wincomplete-uni-patterns -Wredundant-constraints -Wunused-packages
```

lehins commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
```diff
@@ -96,7 +98,8 @@ library testlib
     ghc-options:
         -Wall -Wcompat -Wincomplete-record-updates
-        -Wincomplete-uni-patterns -Wredundant-constraints -Wunused-packages
+        -Wincomplete-uni-patterns -Wredundant-constraints
```
Ditto

lehins commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
```diff
@@ -61,6 +62,7 @@ library
         deepseq,
         FailT,
         formatting,
+        io-classes,
```
Ledger does not need to depend on this package.
```suggestion
```

lehins commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
```diff
@@ -152,7 +156,8 @@ test-suite tests
         -Wall -Wcompat -Wincomplete-record-updates
-        -Wincomplete-uni-patterns -Wredundant-constraints -Wunused-packages
+        -Wincomplete-uni-patterns -Wredundant-constraints
+    -- -Wunused-packages
```
Ditto

lehins commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
```diff
@@ -186,7 +191,8 @@ benchmark bench
         -Wall -Wcompat -Wincomplete-record-updates
-        -Wincomplete-uni-patterns -Wredundant-constraints -Wunused-packages
+        -Wincomplete-uni-patterns -Wredundant-constraints
+    -- -Wunused-packages
```
Ditto

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Crypto.hs:
```diff
@@ -16,6 +16,8 @@ module Cardano.Ledger.Binary.Crypto (
   encodeSignKeyKES,
   decodeSignKeyKES,
+  encodeUnsoundPureSignKeyKES,
+  decodeUnsoundPureSignKeyKES,
```
Remove these as well
```suggestion
```

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Crypto.hs:
```diff
@@ -85,14 +90,25 @@ decodeVerKeyKES :: C.KESAlgorithm v => Decoder s (C.VerKeyKES v)
+decodeUnsoundPureSignKeyKES :: C.UnsoundPureKESAlgorithm v => Decoder s (C.UnsoundPureSignKeyKES v)
+decodeUnsoundPureSignKeyKES = fromPlainDecoder C.decodeUnsoundPureSignKeyKES
+{-# INLINE decodeUnsoundPureSignKeyKES #-}
+
```
We don't need these
```suggestion
```

lehins commented on libs/cardano-ledger-binary/test/Test/Cardano/Ledger/Binary/RoundTripSpec.hs:
```diff
@@ -77,6 +97,30 @@ import Test.Cardano.Ledger.Binary.Arbitrary ()
+         -> IO (SigKES v)
+mkSigKES (seedPSB, msg) =
+  withSK seedPSB $ \sk -> (signKES () 0 (BS.pack msg) sk)
+
```
All KES tests can be removed. Again. We don't need them, if we do not provide those serialization instances

tdammers replied:
I believe we do need ser/deser on verification keys, because that is used to ser/deser OCerts. And in order to generate a verification key for testing, we need to derive it from a signing key, so we need these helpers here.

lehins replied:
I think you missed my point. We need to remove all this new KES keys serialization related functionality from `cardano-ledger-binary`, because ledger does not rely on it.  The fact that OCert contains KES does not interfere with that point, because the `cardano-protocol-tpraos` is really a consensus package. We hope to move it some day into the consensus repo. We can continue using the serialization mechanism provided by `cardano-crypto-class` and `cardano-binary` in those places where KES serialization is needed.

In places like `OCert` we can rely on serialization provided by unsound serialization, it just doesn't need to come from `cardano-ledger-binary`. All you need is `fromPlainDecoder` to lift the regular `Decoder` from `cborg` to the `Decoder` provided by the `cardano-ledger-library` package.

https://github.com/IntersectMBO/cardano-ledger/blob/d24211a6d8a04222a7b9add7a1d9157f2d631933/libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/OCert.hs#L135-L137

The only reason why serialization for KES was provided by `cardano-ledegr-binary` was for completeness. In other words we do not need to provide unsound support, since it will be removed at some point later.

lehins replied:
So, we can continue using sound support for SigKES and for VerKeyKES. While *unsound* support for SignKeyKES remains to be restricted to cardano-base, with some usage of it in ledger, which is restricted to testing.

lehins reviewed (changes requested):
@tdammers Please get rid of any changes related to serialization.
I wish you would have asked me before you went on wasting a lot of time on introducing all those changes. 

We need to get rid of all KES related serialization form ledger, since ledger codebase does not depend on KES at all. So, please revert all of the additions with respect to KES and get rid of any serialization functionality for KES that was provided before

lehins committed:
Add missing CHANGELOG file

neilmayhew commented:
This now passes hydra CI but fails to build locally with `nix develop` and `cabal`. The _Add a workaround the fact plutus-tx-plugin is "special"_ commit makes the local build work but causes a true nix build to fail. We will probably need help from DevX to solve this.

lehins committed:
Add a workaround for `plutus-tx-plugin`

Looks like `plutus-tx-plugin` depends on `ghc` package and it seems that
new dependencies brought in with `cardano-crypto-class-2.2` create a
conflict due to `stm` version used with `ghc-9.6` being incompatible.

lehins committed:
update srp

lehins committed:
Adjust plutus tx config

lehins committed:
Update CHaP to bring in new crypto-class

lehins reviewed (approved):
Now that cardano-crypto-class-2.2 is on CHaP we can finally merge this PR.
